File: app/src/main/java/dev/ragnarok/fenrir/fragment/AudiosFragment.java
package dev.ragnarok.fenrir.fragment;

import static dev.ragnarok.fenrir.util.Objects.nonNull;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.HapticFeedbackConstants;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.work.OneTimeWorkRequest;
import androidx.work.WorkContinuation;
import androidx.work.WorkManager;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.ActivityUtils;
import dev.ragnarok.fenrir.adapter.AudioRecyclerAdapter;
import dev.ragnarok.fenrir.adapter.horizontal.HorizontalPlaylistAdapter;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener;
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback;
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener;
import dev.ragnarok.fenrir.model.Audio;
import dev.ragnarok.fenrir.model.AudioPlaylist;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.AudiosPresenter;
import dev.ragnarok.fenrir.mvp.presenter.base.AccountDependencyPresenter;
import dev.ragnarok.fenrir.mvp.view.IAudiosView;
import dev.ragnarok.fenrir.place.Place;
import dev.ragnarok.fenrir.place.PlaceFactory;
import dev.ragnarok.fenrir.player.MusicPlaybackController;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.AppPerms;
import dev.ragnarok.fenrir.util.CustomToast;
import dev.ragnarok.fenrir.util.DownloadWorkUtils;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.ViewUtils;
import dev.ragnarok.fenrir.view.MySearchView;

public class AudiosFragment extends BaseMvpFragment<AudiosPresenter, IAudiosView>
        implements IAudiosView, HorizontalPlaylistAdapter.Listener {

    public static final String EXTRA_IN_TABS_CONTAINER = "in_tabs_container";
    public static final String ACTION_SELECT = "AudiosFragment.ACTION_SELECT";
    private final AppPerms.doRequestPermissions requestWritePermission = AppPerms.requestPermissions(this,
            new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.READ_EXTERNAL_STORAGE},
            () -> CustomToast.CreateCustomToast(requireActivity()).showToast(R.string.permission_all_granted_text));
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private AudioRecyclerAdapter mAudioRecyclerAdapter;
    private boolean inTabsContainer;
    private final ItemTouchHelper.SimpleCallback simpleItemTouchCallback = new ItemTouchHelper.SimpleCallback(ItemTouchHelper.UP | ItemTouchHelper.DOWN, ItemTouchHelper.LEFT) {
        @Override
        public boolean onMove(@NonNull RecyclerView recyclerView,
                              @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
            return callPresenter(p -> p.fireItemMoved(mAudioRecyclerAdapter.getItemRawPosition(viewHolder.getBindingAdapterPosition()), mAudioRecyclerAdapter.getItemRawPosition(target.getBindingAdapterPosition())), false);
        }

        @Override
        public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int swipeDir) {
            viewHolder.itemView.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS);
            mAudioRecyclerAdapter.notifyItemChanged(viewHolder.getBindingAdapterPosition());
            callPresenter(p -> p.playAudio(requireActivity(), mAudioRecyclerAdapter.getItemRawPosition(viewHolder.getBindingAdapterPosition())));
        }

        @Override
        public boolean isItemViewSwipeEnabled() {
            return !inTabsContainer;
        }

        @Override
        public boolean isLongPressDragEnabled() {
            return !Settings.get().main().isUse_long_click_download() && callPresenter(p -> p.isMyAudio() && p.isNotSearch(), false);
        }
    };
    private boolean isSelectMode;
    private boolean isSaveMode;
    private View headerPlaylist;
    private HorizontalPlaylistAdapter mPlaylistAdapter;

    public static Bundle buildArgs(int accountId, int ownerId, Integer albumId, String access_key) {
        Bundle args = new Bundle();
        args.putInt(Extra.OWNER_ID, ownerId);
        args.putInt(Extra.ACCOUNT_ID, accountId);
        if (nonNull(albumId)) {
            args.putInt(Extra.ALBUM_ID, albumId);
        }
        if (!Utils.isEmpty(access_key)) {
            args.putString(Extra.ACCESS_KEY, access_key);
        }
        return args;
    }

    public static AudiosFragment newInstance(@NonNull Bundle args) {
        AudiosFragment fragment = new AudiosFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static AudiosFragment newInstance(@NonNull Bundle args, boolean isSelect) {
        args.putBoolean(ACTION_SELECT, isSelect);
        AudiosFragment fragment = new AudiosFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        inTabsContainer = requireArguments().getBoolean(EXTRA_IN_TABS_CONTAINER);
        isSelectMode = requireArguments().getBoolean(ACTION_SELECT);
        isSaveMode = false;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_music_main, container, false);
        Toolbar toolbar = root.findViewById(R.id.toolbar);
        if (!inTabsContainer) {
            toolbar.setVisibility(View.VISIBLE);
            ((AppCompatActivity) requireActivity()).setSupportActionBar(toolbar);
        } else {
            toolbar.setVisibility(View.GONE);
        }

        MySearchView searchView = root.findViewById(R.id.searchview);
        searchView.setRightButtonVisibility(false);
        searchView.setLeftIcon(R.drawable.magnify);
        searchView.setOnQueryTextListener(new MySearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                postPresenterReceive(p -> p.fireSearchRequestChanged(query));
                return false;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                postPresenterReceive(p -> p.fireSearchRequestChanged(newText));
                return false;
            }
        });

        mSwipeRefreshLayout = root.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(() -> callPresenter(AudiosPresenter::fireRefresh));
        ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout);

        RecyclerView recyclerView = root.findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(requireActivity()));
        recyclerView.addOnScrollListener(new PicassoPauseOnScrollListener(Constants.PICASSO_TAG));
        recyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener() {
            @Override
            public void onScrollToLastElement() {
                callPresenter(AudiosPresenter::fireScrollToEnd);
            }
        });

        new ItemTouchHelper(simpleItemTouchCallback).attachToRecyclerView(recyclerView);

        FloatingActionButton save_mode = root.findViewById(R.id.save_mode_button);
        FloatingActionButton Goto = root.findViewById(R.id.goto_button);
        save_mode.setVisibility(isSelectMode ? View.GONE : (Settings.get().other().isAudio_save_mode_button() ? View.VISIBLE : View.GONE));
        save_mode.setOnClickListener(v -> {
            isSaveMode = !isSaveMode;
            Goto.setImageResource(isSaveMode ? R.drawable.check : R.drawable.audio_player);
            save_mode.setImageResource(isSaveMode ? R.drawable.ic_dismiss : R.drawable.save);
            mAudioRecyclerAdapter.toggleSelectMode(isSaveMode);
            callPresenter(AudiosPresenter::fireUpdateSelectMode);
        });

        if (isSelectMode) {
            Goto.setImageResource(R.drawable.check);
            save_mode.setImageResource(R.drawable.ic_dismiss);
        } else {
            Goto.setImageResource(R.drawable.audio_player);
            save_mode.setImageResource(R.drawable.save);
        }

        Goto.setOnLongClickListener(v -> {
            if (!isSelectMode && !isSaveMode) {
                Audio curr = MusicPlaybackController.getCurrentAudio();
                if (curr != null) {
                    PlaceFactory.getPlayerPlace(Settings.get().accounts().getCurrent()).tryOpenWith(requireActivity());
                } else {
                    CustomToast.CreateCustomToast(requireActivity()).showToastError(R.string.null_audio);
                }
            } else {
                callPresenter(AudiosPresenter::fireSelectAll);
            }
            return true;
        });
        Goto.setOnClickListener(v -> {
            if (isSelectMode) {
                Intent intent = new Intent();
                intent.putParcelableArrayListExtra(Extra.ATTACHMENTS, callPresenter(p -> p.getSelected(false), new ArrayList<>()));
                requireActivity().setResult(Activity.RESULT_OK, intent);
                requireActivity().finish();
            } else {
                if (isSaveMode) {
                    List<Audio> tracks = callPresenter(p -> p.getSelected(true), new ArrayList<>());
                    isSaveMode = false;
                    Goto.setImageResource(R.drawable.audio_player);
                    save_mode.setImageResource(R.drawable.save);
                    mAudioRecyclerAdapter.toggleSelectMode(isSaveMode);
                    callPresenter(AudiosPresenter::fireUpdateSelectMode);

                    if (!Utils.isEmpty(tracks)) {
                        DownloadWorkUtils.CheckDirectory(Settings.get().other().getMusicDir());
                        int account_id = callPresenter(AccountDependencyPresenter::getAccountId, Settings.get().accounts().getCurrent());
                        WorkContinuation object = WorkManager.getInstance(requireActivity()).beginWith(DownloadWorkUtils.makeDownloadRequestAudio(tracks.get(0), account_id));
                        if (tracks.size() > 1) {
                            List<OneTimeWorkRequest> Requests = new ArrayList<>(tracks.size() - 1);
                            boolean is_first = true;
                            for (Audio i : tracks) {
                                if (is_first) {
                                    is_first = false;
                                    continue;
                                }
                                Requests.add(DownloadWorkUtils.makeDownloadRequestAudio(i, account_id));
                            }
                            object = object.then(Requests);
                        }
                        object.enqueue();
                    }
                } else {
                    Audio curr = MusicPlaybackController.getCurrentAudio();
                    if (curr != null) {
                        int index = callPresenter(p -> p.getAudioPos(curr), -1);
                        if (index >= 0) {
                            recyclerView.scrollToPosition(index + mAudioRecyclerAdapter.getHeadersCount());
                        } else
                            CustomToast.CreateCustomToast(requireActivity()).showToast(R.string.audio_not_found);
                    } else
                        CustomToast.CreateCustomToast(requireActivity()).showToastError(R.string.null_audio);
                }
            }
        });

        mAudioRecyclerAdapter = new AudioRecyclerAdapter(requireActivity(), Collections.emptyList(), callPresenter(AudiosPresenter::isMyAudio, false), isSelectMode, 0, callPresenter(AudiosPresenter::getPlaylistId, null));

        headerPlaylist = inflater.inflate(R.layout.header_audio_playlist, recyclerView, false);
        RecyclerView headerPlaylistRecyclerView = headerPlaylist.findViewById(R.id.header_audio_playlist);
        headerPlaylistRecyclerView.setLayoutManager(new LinearLayoutManager(requireActivity(), LinearLayoutManager.HORIZONTAL, false));
        mPlaylistAdapter = new HorizontalPlaylistAdapter(Collections.emptyList());
        mPlaylistAdapter.setListener(this);
        headerPlaylistRecyclerView.setAdapter(mPlaylistAdapter);

        mAudioRecyclerAdapter.setClickListener(new AudioRecyclerAdapter.ClickListener() {
            @Override
            public void onClick(int position, int catalog, Audio audio) {
                callPresenter(p -> p.playAudio(requireActivity(), position));
            }

            @Override
            public void onEdit(int position, Audio audio) {
                callPresenter(p -> p.fireEditTrackIn(requireActivity(), audio));
            }

            @Override
            public void onDelete(int position) {
                callPresenter(p -> p.fireDelete(position));
            }

            @Override
            public void onUrlPhotoOpen(@NonNull String url, @NonNull String prefix, @NonNull String photo_prefix) {
                PlaceFactory.getSingleURLPhotoPlace(url, prefix, photo_prefix).tryOpenWith(requireActivity());
            }

            @Override
            public void onRequestWritePermissions() {
                requestWritePermission.launch();
            }
        });

        recyclerView.setAdapter(mAudioRecyclerAdapter);
        return root;
    }

    @Override
    public void updatePlaylists(List<AudioPlaylist> playlists) {
        if (nonNull(mPlaylistAdapter)) {
            mPlaylistAdapter.setItems(playlists);
            mPlaylistAdapter.notifyDataSetChanged();
            mAudioRecyclerAdapter.addHeader(headerPlaylist);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        if (!inTabsContainer) {
            Settings.get().ui().notifyPlaceResumed(Place.AUDIOS);
            ActionBar actionBar = ActivityUtils.supportToolbarFor(this);
            if (actionBar != null) {
                actionBar.setTitle(R.string.music);
                actionBar.setSubtitle(null);
            }

            if (requireActivity() instanceof OnSectionResumeCallback) {
                ((OnSectionResumeCallback) requireActivity()).onSectionResume(AbsNavigationFragment.SECTION_ITEM_AUDIOS);
            }

            new ActivityFeatures.Builder()
                    .begin()
                    .setHideNavigationMenu(false)
                    .setBarsColored(requireActivity(), true)
                    .build()
                    .apply(requireActivity());
        }
    }

    @NonNull
    @Override
    public IPresenterFactory<AudiosPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        String accessKey = requireArguments().containsKey(Extra.ACCESS_KEY)
                ? requireArguments().getString(Extra.ACCESS_KEY) : null;
        Integer albumId = requireArguments().containsKey(Extra.ALBUM_ID)
                ? requireArguments().getInt(Extra.ALBUM_ID) : null;
        return () -> new AudiosPresenter(
                requireArguments().getInt(Extra.ACCOUNT_ID),
                requireArguments().getInt(Extra.OWNER_ID),
                albumId,
                accessKey,
                requireArguments().getBoolean(ACTION_SELECT),
                saveInstanceState
        );
    }

    @Override
    public void displayList(List<Audio> audios) {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.setData(audios);
        }
    }

    @Override
    public void notifyListChanged() {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.notifyDataSetChanged();
        }
    }

    @Override
    public void notifyItemMoved(int fromPosition, int toPosition) {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.notifyItemBindableMoved(fromPosition, toPosition);
        }
    }

    @Override
    public void notifyItemRemoved(int index) {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.notifyItemBindableRemoved(index);
        }
    }

    @Override
    public void notifyItemChanged(int index) {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.notifyItemBindableChanged(index);
        }
    }

    @Override
    public void notifyDataAdded(int position, int count) {
        if (nonNull(mAudioRecyclerAdapter)) {
            mAudioRecyclerAdapter.notifyItemBindableRangeInserted(position, count);
        }
    }

    @Override
    public void displayRefreshing(boolean refreshing) {
        if (nonNull(mSwipeRefreshLayout)) {
            mSwipeRefreshLayout.setRefreshing(refreshing);
        }
    }

    @Override
    public void onPlayListClick(AudioPlaylist item, int pos) {
        if (item.getOwnerId() == Settings.get().accounts().getCurrent())
            callPresenter(p -> p.onDelete(item));
        else
            callPresenter(p -> p.onAdd(item));
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/AudiosFragment.kt
package dev.ragnarok.fenrir.fragment

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.HapticFeedbackConstants
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.annotation.StringRes
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import androidx.work.OneTimeWorkRequest
import androidx.work.WorkManager
import com.google.android.material.floatingactionbutton.FloatingActionButton
import com.google.android.material.snackbar.BaseTransientBottomBar
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.Extra
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.ActivityUtils.supportToolbarFor
import dev.ragnarok.fenrir.activity.SendAttachmentsActivity.Companion.startForSendAttachments
import dev.ragnarok.fenrir.adapter.AudioRecyclerAdapter
import dev.ragnarok.fenrir.adapter.horizontal.HorizontalPlaylistAdapter
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener
import dev.ragnarok.fenrir.media.music.MusicPlaybackController.currentAudio
import dev.ragnarok.fenrir.model.Audio
import dev.ragnarok.fenrir.model.AudioPlaylist
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory
import dev.ragnarok.fenrir.mvp.presenter.AudiosPresenter
import dev.ragnarok.fenrir.mvp.view.IAudiosView
import dev.ragnarok.fenrir.place.Place
import dev.ragnarok.fenrir.place.PlaceFactory.getPlayerPlace
import dev.ragnarok.fenrir.place.PlaceFactory.getSingleURLPhotoPlace
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.AppPerms.requestPermissionsAbs
import dev.ragnarok.fenrir.util.DownloadWorkUtils.CheckDirectory
import dev.ragnarok.fenrir.util.DownloadWorkUtils.makeDownloadRequestAudio
import dev.ragnarok.fenrir.util.HelperSimple
import dev.ragnarok.fenrir.util.HelperSimple.needHelp
import dev.ragnarok.fenrir.util.ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme
import dev.ragnarok.fenrir.util.toast.CustomSnackbars
import dev.ragnarok.fenrir.util.toast.CustomToast.Companion.createCustomToast
import dev.ragnarok.fenrir.view.MySearchView

class AudiosFragment : BaseMvpFragment<AudiosPresenter, IAudiosView>(), IAudiosView,
    HorizontalPlaylistAdapter.Listener {
    private val requestWritePermission = requestPermissionsAbs(
        arrayOf(
            Manifest.permission.WRITE_EXTERNAL_STORAGE,
            Manifest.permission.READ_EXTERNAL_STORAGE
        )
    ) {
        createCustomToast(requireActivity()).showToast(R.string.permission_all_granted_text)
    }
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var mAudioRecyclerAdapter: AudioRecyclerAdapter? = null
    private var inTabsContainer = false
    private val simpleItemTouchCallback: ItemTouchHelper.SimpleCallback = object :
        ItemTouchHelper.SimpleCallback(
            ItemTouchHelper.UP or ItemTouchHelper.DOWN,
            ItemTouchHelper.LEFT
        ) {
        override fun onMove(
            recyclerView: RecyclerView,
            viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder
        ): Boolean {
            return presenter?.fireItemMoved(
                mAudioRecyclerAdapter?.getItemRawPosition(viewHolder.bindingAdapterPosition) ?: 0,
                mAudioRecyclerAdapter?.getItemRawPosition(target.bindingAdapterPosition) ?: 0
            ) ?: false
        }

        override fun onSwiped(viewHolder: RecyclerView.ViewHolder, swipeDir: Int) {
            viewHolder.itemView.performHapticFeedback(HapticFeedbackConstants.LONG_PRESS)
            mAudioRecyclerAdapter?.notifyItemChanged(viewHolder.bindingAdapterPosition)
            presenter?.playAudio(
                requireActivity(),
                mAudioRecyclerAdapter?.getItemRawPosition(viewHolder.bindingAdapterPosition) ?: 0
            )
        }

        override fun isItemViewSwipeEnabled(): Boolean {
            return !inTabsContainer
        }

        override fun isLongPressDragEnabled(): Boolean {
            return !Settings.get()
                .main().isUse_long_click_download && presenter?.isMyAudio ?: false && presenter?.isNotSearch ?: false
        }
    }
    private var isSelectMode = false
    private var isSaveMode = false
    private var headerPlaylist: View? = null
    private var mPlaylistAdapter: HorizontalPlaylistAdapter? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        inTabsContainer = requireArguments().getBoolean(EXTRA_IN_TABS_CONTAINER)
        isSelectMode = requireArguments().getBoolean(ACTION_SELECT)
        isSaveMode = false
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_music_main, container, false)
        val toolbar: Toolbar = root.findViewById(R.id.toolbar)
        if (!inTabsContainer) {
            toolbar.visibility = View.VISIBLE
            (requireActivity() as AppCompatActivity).setSupportActionBar(toolbar)
        } else {
            toolbar.visibility = View.GONE
        }
        val searchView: MySearchView = root.findViewById(R.id.searchview)
        searchView.setRightButtonVisibility(false)
        searchView.setLeftIcon(R.drawable.magnify)
        searchView.setOnQueryTextListener(object : MySearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String?): Boolean {
                lazyPresenter {
                    fireSearchRequestChanged(query)
                }
                return false
            }

            override fun onQueryTextChange(newText: String?): Boolean {
                lazyPresenter {
                    fireSearchRequestChanged(newText)
                }
                return false
            }
        })
        mSwipeRefreshLayout = root.findViewById(R.id.refresh)
        mSwipeRefreshLayout?.setOnRefreshListener {
            presenter?.fireRefresh()
        }
        setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout)
        val recyclerView: RecyclerView = root.findViewById(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(requireActivity())
        recyclerView.addOnScrollListener(PicassoPauseOnScrollListener(Constants.PICASSO_TAG))
        recyclerView.addOnScrollListener(object : EndlessRecyclerOnScrollListener() {
            override fun onScrollToLastElement() {
                presenter?.fireScrollToEnd()
            }
        })
        ItemTouchHelper(simpleItemTouchCallback).attachToRecyclerView(recyclerView)
        val save_mode: FloatingActionButton = root.findViewById(R.id.save_mode_button)
        val Goto: FloatingActionButton = root.findViewById(R.id.goto_button)
        save_mode.visibility = when {
            isSelectMode -> View.GONE
            Settings.get()
                .other().isAudio_save_mode_button -> View.VISIBLE
            else -> View.GONE
        }
        save_mode.setOnClickListener {
            isSaveMode = !isSaveMode
            Goto.setImageResource(if (isSaveMode) R.drawable.check else R.drawable.audio_player)
            save_mode.setImageResource(if (isSaveMode) R.drawable.ic_dismiss else R.drawable.save)
            mAudioRecyclerAdapter?.toggleSelectMode(isSaveMode)
            presenter?.fireUpdateSelectMode()
        }
        if (isSelectMode) {
            Goto.setImageResource(R.drawable.check)
            save_mode.setImageResource(R.drawable.ic_dismiss)
        } else {
            Goto.setImageResource(R.drawable.audio_player)
            save_mode.setImageResource(R.drawable.save)
        }
        Goto.setOnLongClickListener {
            if (!isSelectMode && !isSaveMode) {
                val curr = currentAudio
                if (curr != null) {
                    getPlayerPlace(Settings.get().accounts().current).tryOpenWith(requireActivity())
                } else {
                    createCustomToast(requireActivity()).showToastError(R.string.null_audio)
                }
            } else {
                presenter?.fireSelectAll()
            }
            true
        }
        Goto.setOnClickListener {
            if (isSelectMode) {
                val intent = Intent()
                intent.putParcelableArrayListExtra(
                    Extra.ATTACHMENTS, presenter?.getSelected(false) ?: ArrayList()
                )
                requireActivity().setResult(Activity.RESULT_OK, intent)
                requireActivity().finish()
            } else {
                if (isSaveMode) {
                    val tracks: List<Audio> = presenter?.getSelected(true) ?: ArrayList()
                    isSaveMode = false
                    Goto.setImageResource(R.drawable.audio_player)
                    save_mode.setImageResource(R.drawable.save)
                    mAudioRecyclerAdapter?.toggleSelectMode(isSaveMode)
                    presenter?.fireUpdateSelectMode()
                    if (tracks.isNotEmpty()) {
                        CheckDirectory(Settings.get().other().musicDir)
                        val account_id = presenter?.accountId ?: Settings.get().accounts().current
                        var `object` = WorkManager.getInstance(requireActivity()).beginWith(
                            makeDownloadRequestAudio(
                                tracks[0], account_id
                            )
                        )
                        if (tracks.size > 1) {
                            val Requests: MutableList<OneTimeWorkRequest> =
                                ArrayList(tracks.size - 1)
                            var is_first = true
                            for (i in tracks) {
                                if (is_first) {
                                    is_first = false
                                    continue
                                }
                                Requests.add(makeDownloadRequestAudio(i, account_id))
                            }
                            `object` = `object`.then(Requests)
                        }
                        `object`.enqueue()
                    }
                } else {
                    val curr = currentAudio
                    if (curr != null) {
                        val index =
                            presenter?.getAudioPos(curr) ?: -1
                        if (index >= 0) {
                            recyclerView.scrollToPosition(
                                index + (mAudioRecyclerAdapter?.headersCount ?: 0)
                            )
                        } else createCustomToast(requireActivity()).showToast(R.string.audio_not_found)
                    } else createCustomToast(requireActivity()).showToastError(R.string.null_audio)
                }
            }
        }
        mAudioRecyclerAdapter = AudioRecyclerAdapter(
            requireActivity(),
            mutableListOf(),
            presenter?.isMyAudio ?: false,
            isSelectMode,
            0,
            presenter?.playlistId
        )
        headerPlaylist = inflater.inflate(R.layout.header_audio_playlist, recyclerView, false)
        val headerPlaylistRecyclerView: RecyclerView? =
            headerPlaylist?.findViewById(R.id.header_audio_playlist)
        headerPlaylistRecyclerView?.layoutManager =
            LinearLayoutManager(requireActivity(), LinearLayoutManager.HORIZONTAL, false)
        mPlaylistAdapter = HorizontalPlaylistAdapter(mutableListOf())
        mPlaylistAdapter?.setListener(this)
        headerPlaylistRecyclerView?.adapter = mPlaylistAdapter
        mAudioRecyclerAdapter?.setClickListener(object : AudioRecyclerAdapter.ClickListener {
            override fun onClick(position: Int, catalog: Int, audio: Audio) {
                presenter?.playAudio(
                    requireActivity(),
                    position
                )
            }

            override fun onEdit(position: Int, audio: Audio) {
                presenter?.fireEditTrackIn(
                    requireActivity(),
                    audio
                )
            }

            override fun onDelete(position: Int) {
                presenter?.fireDelete(
                    position
                )
            }

            override fun onUrlPhotoOpen(url: String, prefix: String, photo_prefix: String) {
                getSingleURLPhotoPlace(url, prefix, photo_prefix).tryOpenWith(requireActivity())
            }

            override fun onRequestWritePermissions() {
                requestWritePermission.launch()
            }
        })
        recyclerView.adapter = mAudioRecyclerAdapter
        return root
    }

    override fun updatePlaylists(playlist: MutableList<AudioPlaylist>) {
        mPlaylistAdapter?.setItems(playlist)
        mPlaylistAdapter?.notifyDataSetChanged()
        headerPlaylist?.let { mAudioRecyclerAdapter?.addHeader(it) }
    }

    private fun showSnackbar(@StringRes res: Int, isLong: Boolean) {
        CustomSnackbars.createCustomSnackbars(view)
            ?.setDurationSnack(if (isLong) BaseTransientBottomBar.LENGTH_LONG else BaseTransientBottomBar.LENGTH_SHORT)
            ?.defaultSnack(res)?.show()
    }

    override fun showAudioDeadHelper() {
        if (needHelp(HelperSimple.AUDIO_DEAD, 1)) {
            showSnackbar(R.string.audio_dead_helper, true)
        }
    }

    override fun onResume() {
        super.onResume()
        if (!inTabsContainer) {
            Settings.get().ui().notifyPlaceResumed(Place.AUDIOS)
            val actionBar = supportToolbarFor(this)
            if (actionBar != null) {
                actionBar.setTitle(R.string.music)
                actionBar.subtitle = null
            }
            if (requireActivity() is OnSectionResumeCallback) {
                (requireActivity() as OnSectionResumeCallback).onSectionResume(AbsNavigationFragment.SECTION_ITEM_AUDIOS)
            }
            ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity())
        }
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<AudiosPresenter> {
        val accessKey =
            if (requireArguments().containsKey(Extra.ACCESS_KEY)) requireArguments().getString(
                Extra.ACCESS_KEY
            ) else null
        val albumId = if (requireArguments().containsKey(Extra.ALBUM_ID)) requireArguments().getInt(
            Extra.ALBUM_ID
        ) else null
        return object : IPresenterFactory<AudiosPresenter> {
            override fun create(): AudiosPresenter {
                return AudiosPresenter(
                    requireArguments().getInt(Extra.ACCOUNT_ID),
                    requireArguments().getInt(Extra.OWNER_ID),
                    albumId,
                    accessKey,
                    requireArguments().getBoolean(ACTION_SELECT),
                    saveInstanceState
                )
            }
        }
    }

    override fun displayList(audios: MutableList<Audio>) {
        mAudioRecyclerAdapter?.setData(audios)
    }

    override fun notifyListChanged() {
        mAudioRecyclerAdapter?.notifyDataSetChanged()
    }

    override fun notifyItemMoved(fromPosition: Int, toPosition: Int) {
        mAudioRecyclerAdapter?.notifyItemBindableMoved(fromPosition, toPosition)
    }

    override fun notifyItemRemoved(index: Int) {
        mAudioRecyclerAdapter?.notifyItemBindableRemoved(index)
    }

    override fun notifyItemChanged(index: Int) {
        mAudioRecyclerAdapter?.notifyItemBindableChanged(index)
    }

    override fun notifyDataAdded(position: Int, count: Int) {
        mAudioRecyclerAdapter?.notifyItemBindableRangeInserted(position, count)
    }

    override fun displayRefreshing(refreshing: Boolean) {
        mSwipeRefreshLayout?.isRefreshing = refreshing
    }

    override fun onPlayListClick(item: AudioPlaylist, pos: Int) {
        if (item.getOwnerId() == Settings.get()
                .accounts().current
        ) presenter?.onDelete(item) else presenter?.onAdd(item)
    }

    override fun onShareClick(item: AudioPlaylist, pos: Int) {
        startForSendAttachments(requireActivity(), Settings.get().accounts().current, item)
    }

    companion object {
        const val EXTRA_IN_TABS_CONTAINER = "in_tabs_container"
        const val ACTION_SELECT = "AudiosFragment.ACTION_SELECT"
        fun buildArgs(accountId: Int, ownerId: Int, albumId: Int?, access_key: String?): Bundle {
            val args = Bundle()
            args.putInt(Extra.OWNER_ID, ownerId)
            args.putInt(Extra.ACCOUNT_ID, accountId)
            if (albumId != null) {
                args.putInt(Extra.ALBUM_ID, albumId)
            }
            if (!access_key.isNullOrEmpty()) {
                args.putString(Extra.ACCESS_KEY, access_key)
            }
            return args
        }

        fun newInstance(args: Bundle): AudiosFragment {
            val fragment = AudiosFragment()
            fragment.arguments = args
            return fragment
        }

        fun newInstance(args: Bundle, isSelect: Boolean): AudiosFragment {
            args.putBoolean(ACTION_SELECT, isSelect)
            val fragment = AudiosFragment()
            fragment.arguments = args
            return fragment
        }
    }
}